crates/ibc/src/clients/ics07_tendermint/client_state.rs

```diff
@@ -876,13 +879,145 @@ impl Ics2ClientState for ClientState {
+        let cons_upgrade_merkle_path = MerklePath {
+            key_path: cons_upgrade_path,
+        };
+
+        let cons_state_value = Protobuf::<RawTmConsensusState>::encode_vec(&upgraded_tm_cons_state)
+            .map_err(ClientError::Encode)?;
+
+        // Verify the proof of the upgraded consensus state
+        merkle_proof_upgrade_cons_state
+            .verify_membership(
+                &self.proof_specs,
+                root.clone().into(),
+                cons_upgrade_merkle_path,
+                cons_state_value,
+                0,
+            )
+            .map_err(ClientError::Ics23Verification)?;
+
+        Ok(())
+    }
+
+    fn update_state_with_upgrade_client(
         &self,
-        _consensus_state: Any,
-        _proof_upgrade_client: RawMerkleProof,
-        _proof_upgrade_consensus_state: RawMerkleProof,
+        upgraded_client_state: Any,
+        upgraded_consensus_state: Any,
     ) -> Result<UpdatedState, ClientError> {
-        unimplemented!()
+        let upgraded_tm_client_state = TmClientState::try_from(upgraded_client_state)?;
+        let upgraded_tm_cons_state = TmConsensusState::try_from(upgraded_consensus_state)?;
+
+        // Construct new client state and consensus state
+        // Relayer chosen client parameters are ignored.
+        // All chain-chosen parameters come from committed client, all client-chosen parameters
+        // come from current client.
+        let new_client_state = TmClientState::new(
+            upgraded_tm_client_state.chain_id,
+            self.trust_level,
+            self.trusting_period,
+            upgraded_tm_client_state.unbonding_period,
+            self.max_clock_drift,
+            upgraded_tm_client_state.latest_height,
+            upgraded_tm_client_state.proof_specs,
+            upgraded_tm_client_state.upgrade_path,
+            upgraded_tm_client_state.allow_update,
+            upgraded_tm_client_state.frozen_height,
+        )?;
+
+        // The new consensus state is merely used as a trusted kernel against which headers on the new
+        // chain can be verified. The root is just a stand-in sentinel value as it cannot be known in advance, thus no proof verification will pass.
+        // The timestamp and the NextValidatorsHash of the consensus state is the blocktime and NextValidatorsHash
+        // of the last block committed by the old chain. This will allow the first block of the new chain to be verified against
+        // the last validators of the old chain so long as it is submitted within the TrustingPeriod of this client.
+        // NOTE: We do not set processed time for this consensus state since this consensus state should not be used for packet verification
+        // as the root is empty. The next consensus state submitted using update will be usable for packet-verification.
+        let new_consensus_state = TmConsensusState::new(
+            vec![].into(),
```

Comment:
I don't have a deep understanding of how the SDK does upgrades. For me, many questions need to be answered here. Why does it make sense to put an empty root here? What is the behavior of the next `update_client` when it sees this root? Is the host expected to do anything else (that would normally be provided by the SDK)? Does the SDK's strategy for chain upgrades make sense for ibc-rs, given that our users typically have radically different architectures from the SDK?

This set of questions is why I [initially suggested](https://github.com/cosmos/ibc-rs/issues/141#issuecomment-1349633869) we simply output an error for now, and figure out the details later. However, maybe you have a better understanding of this than I do at this point? If so, then perhaps we'd need to write an ADR about this to explain why we designed our upgrade API like so, and answer the questions I asked above.

WDYT?

Reply:
Actually, I followed very similar logic to what has been implemented on the `ibc-go` with some changes to make it fit our context. It, of course, is largely compatible with SDK chains, but I think most of the validation checkings (even all) and some of the update-related stuff apply also here. 
But, I understand what you mean. there are some parts (as you truly mentioned) need further investigation. 
If the decision would be to postpone the implementation until we have some real cases, that's a different story. Though, I propose that even in its minimal shape it makes sense to have a kind of implementation as I think future cases may differ considerably...
Thus, if we agree on its existence, I will work on the ADR. We can at least come up with some clear explanations

Reply:
> If the decision would be to postpone the implementation until we have some real cases, that's a different story. Though, I propose that, even in its minimal shape, to have a kind of implementation as I think future cases may differ considerably...

I'm okay with having an implementation right now, as long as it is self-consistent and well-documented. Which is the perfect use case for an ADR.

> Thus, if we agree on its existence, I will work on the ADR. We can at least come up with some clear explanations

Yep, if you're interested, then please write an ADR! We can discuss the specifics there.

Reply:
Thanks for bringing this up. 
Alright, I'll go for it 👍

Reply:
Issue #362 - PR #383